OpenRA.Mods.Common/Traits/Buildings/ProductionAirdrop.cs

```diff
@@ -112,11 +103,142 @@ public override bool Produce(Actor self, ActorInfo producee, string productionTy
+					var loc = self.Location.ToMPos(map);
+					startPos = new MPos(loc.U + map.Bounds.Width, loc.V).ToCPos(map);
+
+					spawnFacing = GetSpawnFacing(self.Location, startPos);
+					break;
+
+				case EntryType.PlayerSpawnClosestEdge:
+					if (mpStart == null)
+						break;
+
+					var spawn = mpStart.Start[owner];
+					var center = new MPos(bounds.Left + bounds.Width / 2, bounds.Top + bounds.Height / 2).ToCPos(map);
+					var spawnVec = spawn - center;
+					startPos = spawn + spawnVec * (Exts.ISqrt((bounds.Height * bounds.Height + bounds.Width * bounds.Width) / (4 * spawnVec.LengthSquared)));
+
+					spawnFacing = GetSpawnFacing(self.Location, startPos);
+					break;
+
+				case EntryType.DropSiteClosestEdge:
+					startPos = self.World.Map.ChooseClosestEdgeCell(self.Location);
+
+					spawnFacing = GetSpawnFacing(self.Location, startPos);
+
+					bounds = self.World.Map.Bounds;
+					if (info.SpawnOffset.X != 0 && startPos.X != bounds.X && startPos.X != bounds.X + bounds.Width)
+						startPos += new CVec(info.SpawnOffset.X, 0);
+
+					if (info.SpawnOffset.Y != 0 && startPos.Y != bounds.Y && startPos.Y != bounds.Y + bounds.Height)
+						startPos += new CVec(0, info.SpawnOffset.Y);
+					break;
+			}
+
+			var exitPositionFunc = GetExitPositionFunc(self, startPos, spawnFacing);
+
+			return new DeliveryActorPathInfo(startPos, spawnFacing, info.LandingFacing, exitPositionFunc);
+		}
+
+		static int GetSpawnFacing(CPos targetLocation, CPos spawnLocation)
+		{
+			var direction = targetLocation - spawnLocation;
+			var spawnDirection = new WVec(direction.X, direction.Y, 0);
+			return spawnDirection.Yaw.Facing;
+		}
+
+		Func<Actor, CPos> GetExitPositionFunc(Actor self, CPos startPos, int spawnFacing)
+		{
+			Func<Actor, CPos> exitPositionFunc = actor => CPos.Zero;
+			switch (info.ExitType)
+			{
+				case ExitType.OppositeToEntry:
+					exitPositionFunc = actor =>
+					{
+						var rotation = WRot.FromFacing(spawnFacing);
+						var delta = new WVec(0, -1024, 0).Rotate(rotation);
+						var finishEdge = actor.CenterPosition + self.World.Map.DistanceToEdge(actor.CenterPosition, delta).Length * delta / 1024;
+						return new CPos(finishEdge.X / 1024, finishEdge.Y / 1024, 0);
+					};
+
+					break;
+
+				case ExitType.StraightAhead:
+					exitPositionFunc = actor =>
+					{
+						var rotation = WRot.FromFacing(actor.Orientation.Yaw.Facing);
+						var delta = new WVec(0, -1024, 0).Rotate(rotation);
+						var finishEdge = actor.CenterPosition + self.World.Map.DistanceToEdge(actor.CenterPosition, delta).Length * delta / 1024;
+						return new CPos(finishEdge.X / 1024, finishEdge.Y / 1024, 0);
+					};
+
+					break;
+
+				case ExitType.SameAsEntry:
+					exitPositionFunc = actor => startPos;
+					break;
+
+				case ExitType.Closest:
+					exitPositionFunc = actor =>
+					{
+						// By the time the func gets executed the map bounds could have changed, so fetching them fresh.
```

Comment:
Can they?

Reply:
Currently no, but that is a prerequisite for TS missions, so I assume we'll get it eventually in time for the TS Public Alpha.

Reply:
That feature should really come with an INotify callback, so I don't think we should calculate this here, or make assumptions about future code at this moment.

Reply:
Fair enough on the `INotify`.